polaris-service/src/main/java/org/apache/polaris/service/persistence/MetadataCacheManager.java

```diff
@@ -0,0 +1,143 @@
+      TableIdentifier tableIdentifier,
+      long maxBytesToCache,
+      PolarisCallContext callContext,
+      PolarisMetaStoreManager metastoreManager,
+      PolarisResolutionManifestCatalogView resolvedEntityView,
+      Supplier<TableMetadata> fallback) {
+    LOGGER.debug(String.format("Loading cached metadata for %s", tableIdentifier));
+    PolarisResolvedPathWrapper resolvedEntities =
+        resolvedEntityView.getResolvedPath(tableIdentifier, PolarisEntitySubType.TABLE);
+    TableLikeEntity tableLikeEntity = TableLikeEntity.of(resolvedEntities.getRawLeafEntity());
+    boolean isCacheValid =
+        tableLikeEntity.getMetadataCacheLocation() != null
+            && tableLikeEntity
+                .getMetadataLocation()
+                .equals(tableLikeEntity.getMetadataCacheLocation());
+    if (isCacheValid) {
+      LOGGER.debug(String.format("Using cached metadata for %s", tableIdentifier));
+      return TableMetadataParser.fromJson(tableLikeEntity.getMetadataCacheContent());
+    } else {
+      TableMetadata metadata = fallback.get();
+      PolarisMetaStoreManager.EntityResult cacheResult =
+          cacheTableMetadata(
+              tableLikeEntity,
+              metadata,
+              maxBytesToCache,
+              callContext,
+              metastoreManager,
+              resolvedEntityView);
+      if (!cacheResult.isSuccess()) {
+        LOGGER.debug(String.format("Failed to cache metadata for %s", tableIdentifier));
+      }
+      return metadata;
+    }
+  }
+
+  /**
+   * Attempt to add table metadata to the cache
+   *
+   * @return The result of trying to cache the metadata
+   */
+  private static PolarisMetaStoreManager.EntityResult cacheTableMetadata(
+      TableLikeEntity tableLikeEntity,
+      TableMetadata metadata,
+      long maxBytesToCache,
+      PolarisCallContext callContext,
+      PolarisMetaStoreManager metaStoreManager,
+      PolarisResolutionManifestCatalogView resolvedEntityView) {
+    String json = TableMetadataParser.toJson(metadata);
+    // We should not reach this method in this case, but check just in case...
+    if (maxBytesToCache != PolarisConfiguration.METADATA_CACHE_MAX_BYTES_NO_CACHING) {
+      long sizeInBytes = json.getBytes(StandardCharsets.UTF_8).length;
+      if (sizeInBytes > maxBytesToCache) {
+        LOGGER.debug(
+            String.format(
+                "Will not cache metadata for %s; metadata is %d bytes and the limit is %d",
+                tableLikeEntity.getTableIdentifier(), sizeInBytes, maxBytesToCache));
+        return new PolarisMetaStoreManager.EntityResult(
+            PolarisMetaStoreManager.ReturnStatus.SUCCESS, null);
+      } else {
+        LOGGER.debug(
+            String.format("Caching metadata for %s", tableLikeEntity.getTableIdentifier()));
+        TableLikeEntity newTableLikeEntity =
+            new TableLikeEntity.Builder(tableLikeEntity)
+                .setMetadataContent(tableLikeEntity.getMetadataLocation(), json)
```

Comment:
Storing JSON uncompressed and as a string is a waste of resources on the network, on heap and in the database.
I'd suggest to use Smile serialization format and likely compress that as well.

Reply:
> Personally, I would rather us use a more structured schema (e.g. a more relational one) that supports partial metadata loading rather than just compress the current JSON-based approach.

Definitely this. However, IMO we should design this approach before we add something else. But I'm not sure whether "relational" is the right terminology here?

I still have concerns about the handling of uncompressed strings and duplicate re-serializations, causing a lot of heap and CPU pressure, especially for the really huge metadata objects.